pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,1017 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bgp
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	discoveryinformers "k8s.io/client-go/informers/discovery/v1"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	discoverylisters "k8s.io/client-go/listers/discovery/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/net"
+	"k8s.io/utils/strings/slices"
+
+	"antrea.io/antrea/pkg/agent/bgp"
+	"antrea.io/antrea/pkg/agent/bgp/gobgp"
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	"antrea.io/antrea/pkg/apis/crd/v1beta1"
+	crdinformersv1a1 "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha1"
+	crdinformersv1b1 "antrea.io/antrea/pkg/client/informers/externalversions/crd/v1beta1"
+	crdlistersv1a1 "antrea.io/antrea/pkg/client/listers/crd/v1alpha1"
+	crdlistersv1b1 "antrea.io/antrea/pkg/client/listers/crd/v1beta1"
+	"antrea.io/antrea/pkg/features"
+	"antrea.io/antrea/pkg/util/env"
+	utilipset "antrea.io/antrea/pkg/util/sets"
+)
+
+const (
+	controllerName = "BGPPolicyController"
+	// How long to wait before retrying the processing of a BGPPolicy change.
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
+	// Default number of workers processing a BGPPolicy change.
+	defaultWorkers = 4
+	// Disable resyncing.
+	resyncPeriod time.Duration = 0
+)
+
+const (
+	bgpRouteIDAnnotation = "antrea.io/bgp-route-id"
+
+	defaultBGPListenPort int32 = 179
+)
+
+const (
+	ipv4Suffix = "/32"
+	ipv6Suffix = "/128"
+)
+
+var (
+	protocolIPv4 = net.IPv4
+	protocolIPv6 = net.IPv6
+
+	newBGPServerFn = func(globalConfig *bgp.GlobalConfig) bgp.Interface {
+		return gobgp.NewGoBGPServer(globalConfig)
+	}
+)
+
+type nodeToBGPPolicyBinding struct {
+	effectiveBP    string
+	alternativeBPs sets.Set[string]
+}
+
+type bgpPolicyState struct {
+	// The local BGP server created for the BGPPolicy.
+	bgpServer bgp.Interface
+	// The port on which local the BGP server listens.
+	listenPort int32
+	// The AS number used by the local BGP server.
+	localASN int32
+	// The router ID used by the local BGP server.
+	routerID string
+	// Routes to be advertised to BGP peers.
+	routes sets.Set[bgp.Route]
+	// peers maps IP families to concatenated strings of BGP peer IP addresses and ASNs.
+	// Example: "192.168.77.100-65000", "2001::1-65000".
+	peerKeys sets.Set[string]
+	// peerConfigs maps concatenated string of BGP peer IP addresses and ASN to the configuration of the peer.
+	peerConfigs map[string]bgp.PeerConfig
```

Comment:
`peerKeys` can be removed. If so, we still need to collect all the keys of the current and the stale `peerConfigs` to two sets, and then get the peers to be added or deleted.

Reply:
It can be easily got by `sets.KeySet(peerConfigs)`, it's not a common "optimization" to use a separate set to store another map's keys, which may not be even more efficient as it still needs to maintain an extra set when writing.